### PR TITLE
out_opentelemetry: support metadata key properties

### PR DESCRIPTION
**Pull request by nokute78**
This patch is to support several metadata key properties to get a OTLP values from metadata.
See also https://github.com/fluent/fluent-bit/pull/8294 and https://github.com/fluent/fluent-bit/issues/8205

|key|Description|Default|
|---|--|---|
|logs_observed_timestamp_metadata_key| Specify an ObservedTimestamp key to look up in the metadata. |ObservedKey|
|logs_timestamp_metadata_key|Specify an Timestamp key to look up in the metadata. |Timestamp|
|logs_severity_key_metadata_key|Specify an SeverityText key to look up in the metadata.|SeverityText|
|logs_severity_number_metadata_key| Specify an SeverityNumber key to look up in the metadata.|SeverityNumber|
|logs_trace_flags_metadata_key|Specify an Flags key to look up in the metadata.|Flags|
|logs_span_id_metadata_key|Specify an SpanId key to look up in the metadata.|SpanId|
|logs_trace_id_metadata_key|Specify an TraceId key to look up in the metadata.|TraceId|
|logs_attributes_metadata_key|Specify an Attributes key to look up in the metadata.|Attributes|

- [X] Timestamp
- [X] ObservedTimestamp
- [X] TraceId
- [X] SpanId
- [X] TraceFlags
- [X] SeverityText
- [X] SeverityNumber
- [ ] Resource
- [ ] InstrumentationScope
- [X] Attributes

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [X] Example configuration file for the change
- [X] Debug log output from testing the change

- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [N/A] Backport to latest stable release.

## Configuration

```
[INPUT]
    NAME dummy
    samples 1
    Dummy {"log":"aaaa"}
    Metadata {"ObservedTimestamp":1234, "Timestamp":3456, "SeverityText":"WARN", "SeverityNumber":5, "TraceFlags":13, "Attributes":{"log.file.name":"a.log"}}

[OUTPUT]
    Name   stdout
    Match *

[OUTPUT]
    Name opentelemetry
    Host localhost
    Port 6969
    Match *
```

Config for otelcol-contrib:
```yaml
receivers:
  otlp:
    protocols:
      http:
        endpoint: localhost:6969

exporters:
  file:
    path: 2_out.json

service:
  telemetry:
    logs:
      level: debug
  pipelines:
    logs:
      receivers: [otlp]
      exporters: [file]
```

## Debug/Valgrind output

```
[2024/02/11 12:41:01] [ info] [fluent bit] version=3.0.0, commit=9652b0dc6f, pid=61472
[2024/02/11 12:41:01] [ info] [storage] ver=1.2.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2024/02/11 12:41:01] [ info] [cmetrics] version=0.6.6
[2024/02/11 12:41:01] [ info] [ctraces ] version=0.4.0
[2024/02/11 12:41:01] [ info] [input:dummy:dummy.0] initializing
[2024/02/11 12:41:01] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
[2024/02/11 12:41:02] [ info] [output:stdout:stdout.0] worker #0 started
[2024/02/11 12:41:02] [ info] [sp] stream processor started
[0] dummy.0: [[1707622862.439011464, {"ObservedTimestamp"=>1234, "Timestamp"=>3456, "SeverityText"=>"WARN", "SeverityNumber"=>5, "TraceFlags"=>13, "Attributes"=>{"log.file.name"=>"a.log"}}], {"log"=>"aaaa"}]
[2024/02/11 12:41:03] [ info] [output:opentelemetry:opentelemetry.1] localhost:6969, HTTP status=200


^C[2024/02/11 12:41:05] [engine] caught signal (SIGINT)
[2024/02/11 12:41:05] [ warn] [engine] service will shutdown in max 5 seconds
[2024/02/11 12:41:05] [ info] [input] pausing dummy.0
[2024/02/11 12:41:05] [ info] [engine] service has stopped (0 pending tasks)
[2024/02/11 12:41:05] [ info] [input] pausing dummy.0
[2024/02/11 12:41:05] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2024/02/11 12:41:05] [ info] [output:stdout:stdout.0] thread worker #0 stopped
==61472== 
==61472== HEAP SUMMARY:
==61472==     in use at exit: 0 bytes in 0 blocks
==61472==   total heap usage: 2,236 allocs, 2,236 frees, 1,002,111 bytes allocated
==61472== 
==61472== All heap blocks were freed -- no leaks are possible
==61472== 
==61472== For lists of detected and suppressed errors, rerun with: -s
```

Output of otelcol-contrib:
```json
{
  "resourceLogs": [
    {
      "resource": {},
      "scopeLogs": [
        {
          "scope": {},
          "logRecords": [
            {
              "timeUnixNano": "3456",
              "observedTimeUnixNano": "1234",
              "severityNumber": 5,
              "severityText": "WARN",
              "body": {
                "kvlistValue": {
                  "values": [
                    {
                      "key": "log",
                      "value": {
                        "stringValue": "aaaa"
                      }
                    }
                  ]
                }
              },
              "attributes": [
                {
                  "key": "log.file.name",
                  "value": {
                    "stringValue": "a.log"
                  }
                }
              ],
              "flags": 13,
              "traceId": "",
              "spanId": ""
            }
          ]
        }
      ]
    }
  ]
}
```
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by nokute78** [comment]
@edsiper Thank you for feedback.

> I would suggest to remove the word metadata from the new config properties

I think following values should be specified.
- Logs/Metrics/Traces
- Field of logs
   -  https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition
- Metadata/Message

I sent a patch to store OTLP data as a `metadata`. https://github.com/fluent/fluent-bit/pull/8294. These properties for the patch.
On the other hand, user may use a value of `message`. 
e.g. OTEL supports `SeverityText` and `SeverityNumber` and user may want to use a value from in_syslog.
https://opentelemetry.io/docs/specs/otel/logs/data-model-appendix/

So I added `metadata` to properties to clearly distinguish `metadata` and `message`.

Or it may be better to support properties to seek a value from `metadata` AND `message`.
e.g. If user set `logs_attributes_key`, out_otel lookup `event.metadata` and `event.body`.

> use record accessor pattern, so the end user gets the flexibility to use any subkey of the record.

I think so. I'll modify this patch.

**Comment by nokute78** [comment]
@edsiper 

> Attributes handling will be handled differently as described in #8491, so it won't be needed for now.

If I wrong please correct me.
I think #8491 will not fix #8205 since it converts other metadata to Attributes.

This PR is to fix https://github.com/fluent/fluent-bit/issues/8205. It is to support forwarding OTLP.
I sent a patch https://github.com/fluent/fluent-bit/pull/8294 and it creates following metadata for forwarding.
```json
{"ObservedTimestamp":1706312272306657295, "Attributes":{"log.file.name":"a.log"}, "TraceFlags":0, "Resource":{"host.name":"taka-VirtualBox", "os.type":"linux"}, "InstrumentationScope":{"Name":"test_scope", "Version":"1"}}
```

https://github.com/fluent/fluent-bit/pull/8491 will store all above metadata as Attributes of OTLP.
It means fluent-bit changes original Attributes and not forward original OTLP fields.

I tested following configuration and https://github.com/fluent/fluent-bit/pull/8491

```
[INPUT]
    name dummy
    Metadata {"ObservedTimestamp":1234, "Timestamp":3456, "SeverityText":"WARN", "SeverityNumber":5, "TraceFlags":13, "Attributes":{"log.file.name":"a.log", "test":"val"}}
    dummy {"name": "bill", "lastname": "gates", "log": {"abc": {"def":123}, "saveme": true}}

[OUTPUT]
    Name   stdout
    Match *

[OUTPUT]
    Name opentelemetry
    Host localhost
    Port 6969
    Match *
    logs_body_key $name
    logs_body_key $log['abc']['def']
```

configuration for otelcol-contrib
```yaml
receivers:
  otlp:
    protocols:
      http:
        endpoint: localhost:6969

exporters:
  file:
    path: 2_out.json

service:
  telemetry:
    logs:
      level: debug
  pipelines:
    logs:
      receivers: [otlp]
      exporters: [file]
```

Output of otel-collector
```json
{
  "resourceLogs": [
    {
      "resource": {},
      "scopeLogs": [
        {
          "scope": {},
          "logRecords": [
            {
              "timeUnixNano": "1708128958966094604",
              "body": {
                "stringValue": "bill"
              },
              "attributes": [
                {
                  "key": "ObservedTimestamp",
                  "value": {
                    "intValue": "1234"
                  }
                },
                {
                  "key": "Timestamp",
                  "value": {
                    "intValue": "3456"
                  }
                },
                {
                  "key": "SeverityText",
                  "value": {
                    "stringValue": "WARN"
                  }
                },
                {
                  "key": "SeverityNumber",
                  "value": {
                    "intValue": "5"
                  }
                },
                {
                  "key": "TraceFlags",
                  "value": {
                    "intValue": "13"
                  }
                },
                {
                  "key": "Attributes",
                  "value": {
                    "kvlistValue": {
                      "values": [
                        {
                          "key": "log.file.name",
                          "value": {
                            "stringValue": "a.log"
                          }
                        },
                        {
                          "key": "test",
                          "value": {
                            "stringValue": "val"
                          }
                        }
                      ]
                    }
                  }
                },
                {
                  "key": "lastname",
                  "value": {
                    "stringValue": "gates"
                  }
                },
                {
                  "key": "log",
                  "value": {
                    "kvlistValue": {
                      "values": [
                        {
                          "key": "abc",
                          "value": {
                            "kvlistValue": {
                              "values": [
                                {
                                  "key": "def",
                                  "value": {
                                    "intValue": "123"
                                  }
                                }
                              ]
                            }
                          }
                        },
                        {
                          "key": "saveme",
                          "value": {
                            "boolValue": true
                          }
                        }
                      ]
                    }
                  }
                }
              ],
              "traceId": "",
              "spanId": ""
            }
          ]
        }
      ]
    }
  ]
}

```

**Comment by edsiper** [comment]
Hi @nokute78 , 

#8491 aims to solve only body and attributes that are set as random keys from the record. Yes, I agree that PR is not to solve everything. 

If we let the concept of forwarding and in_opentelemetry in the side for a moment, I think this PR should purely focus on the discovery and setting of the keys you described above that are part from the data model:

key | Description | Default
-- | -- | --
logs_observed_timestamp_key | Specify an ObservedTimestamp key to look up in the metadata. | ObservedKey
logs_timestamp_key | Specify an Timestamp key to look up in the metadata. | Timestamp
logs_severity_key_key | Specify an SeverityText key to look up in the metadata. | SeverityText
logs_severity_number_key | Specify an SeverityNumber key to look up in the metadata. | SeverityNumber
logs_trace_flags_key | Specify an Flags key to look up in the metadata. | Flags
logs_span_id_key | Specify an SpanId key to look up in the metadata. | SpanId
logs_trace_id_key | Specify an TraceId key to look up in the metadata. | TraceId

I skipped the `logs_attributes_key` since it must have a separate handling. 

Since the urgency is to solve `any structured message` to `OTLP` first, do you think you can adjust the imlementation for the config above ? (by using record accessor as suggested?), on that way we can solve this primary use case, once community confirm is ok we can focus on forwarding `in_opentelemetry` &rarr; `out_opentelemetry`. 

let me know what do you think, thanks

**Comment by nokute78** [comment]
@edsiper  Thank you for comment.

> I skipped the logs_attributes_key since it must have a separate handling.

Could you tell the detail ? Why are `Attributes` special ?

https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-attributes
```
SHOULD follow OpenTelemetry semantic conventions for attributes. This field is optional.
```

`Attributes` is an optional field. It should be valid if `Attributes` is empty.
And spec requests to follow [OTEL semantic conventions](https://opentelemetry.io/docs/specs/semconv/attributes-registry/).
I think #8491 can be against the semantic conventions.

>  the urgency is to solve any structured message to OTLP first

In my opinion, out_opentelemetry should be 

- Default value of optional fields are empty.
   - Note: almost of fields are optional. `Attributes` is one of them.
   - Please see https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition
- `logs_body_key`  is to scrape off `Body` and don't touch `Attributes`.
- `logs_AAA_metadata_key` of this PR to specify a value of optional field. If the value is found, fluent-bit set it as a value of optional field.
   - e.g. User specify `logs_attributes_metadata_key` and the value is found, fluent-bit overwrites `Attributes` of OTLP.
- It may be better to set default value of optional field.
   - e.g. A property `default_trace_id` and user can set the default trace_id.


> logs_observed_timestamp_metadata_key

By the way, this `metadata` means a `METADATA` of fluent-bit event. That is not for `MESSAGE` a.k.a record.
https://docs.fluentbit.io/manual/concepts/key-concepts#event-format
```
[[TIMESTAMP, METADATA], MESSAGE]
             ^^^^^^^^
```

**Comment by edsiper** [comment]
@nokute78 I have updated #8491, now it does not auto-populate attributes with remaining keys unless `logs_body_key_attributes` is enabled (disabled by default). Note that PR already take care of log metadata and set them as attributes. So I think this PR should focus in the other keys as mentioned.

**Comment by nokute78** [comment]
@edsiper #8491 is for `MESSAGE` and this patch is for `METADATA`. They are different feature.

By the way I think it is better that the default value of `logs_*_metadata_key` is NULL to disable by default.

**Comment by nokute78** [comment]
@braydonk Could you check this PR ?
This patch is to set a value for OTLP from metadata.

As you know https://github.com/fluent/fluent-bit/pull/8294 is to store OTLP field as metadata of fluent-bit event.
This patch is to read metadata and set these value as fields of OTLP.
It is to forward OTLP using fluent-bit. https://github.com/fluent/fluent-bit/issues/8205

**Comment by cb645j** [comment]
@nokute78 @edsiper When will this be merged? And which release will it be part of?

File: plugins/out_opentelemetry/opentelemetry.c
Comment (braydonk) [inline plugins/out_opentelemetry/opentelemetry.c:1505]: This and a number of other descriptions should be `a` instead of `an` when the word following it starts with a consonant.

File: plugins/out_opentelemetry/opentelemetry.c
Comment (braydonk) [inline plugins/out_opentelemetry/opentelemetry.c:1500]: Since these are using record accessors now, is `_key` still the right word to use for the config field name?

**Review by braydonk — COMMENTED** [review]
It does not seem to be exactly related to https://github.com/fluent/fluent-bit/issues/8205 since that was about the Resource and Scope attribute stuff but this is still important for a complete OTLP Log solution.

Looks good to me on an individual record level.

**Comment by cb645j** [comment]
@braydonk @edsiper @nokute78 how can we get this completed do we can merge.  Fluent bit will not work correctly with opentelemtry until this is merged. My company is blocked by these fluentbit otel issues.

File: plugins/out_opentelemetry/opentelemetry.c
Comment (nokute78) [reply plugins/out_opentelemetry/opentelemetry.c:1500]: I think it is right word.

Some plugins support a record accessor as a `key`.
e.g. https://docs.fluentbit.io/manual/pipeline/filters/rewrite-tag#key

It allows that user doesn't need to take care that a key is nested or not.

File: plugins/out_opentelemetry/opentelemetry.c
Comment (nokute78) [reply plugins/out_opentelemetry/opentelemetry.c]: Oops. Thank you for pointing it.
I fixed them.
https://github.com/fluent/fluent-bit/pull/8475/commits/7660e3f25d4bf0520878b3c1dc6368b514a90a32

**Comment by cb645j** [comment]
@braydonk can you approve please

**Comment by braydonk** [comment]
My approval won't have any effect on moving this PR forward as I'm not a maintainer and my review wasn't directly requested, but I can press the button if it helps.

**Review by braydonk — APPROVED** [review]

**Comment by cb645j** [comment]
> My approval won't have any effect on moving this PR forward as I'm not a maintainer and my review wasn't directly requested, but I can press the button if it helps.

I see, i guess we need someone with the magic touch

**Comment by cb645j** [comment]
@edsiper @koleini @fujimotos @leonardo-albertovich Can you please approve/review so we can move this forward.

**Comment by cb645j** [comment]
@nokute78 yes, i understand. The one for message was already merged by @edsiper.  However it seems we are still unable to set the toplevel traceid and spanid properties/fields which i what i need.  To my understanding this PR address it,

I think the resource property can be done separately

**Comment by cb645j** [comment]
@nokute78 I see what your saying now.  Can we not change this to get this data from the message? Im not sure how you even set the metadata.... All my log events have a blank metadata.

**Comment by cb645j** [comment]
@nokute78 I will open up a separate PR

**Comment by cb645j** [comment]
@nokute78 @edsiper I tested this and the trace id and span id fields do not work as strings.  The other fields work but trace id and span id dont. It looks like @nokute78 test above did not cover testing these fields.  In his metadat example test above he left these out. This is what he had. 

Metadata {"ObservedTimestamp":1234, "Timestamp":3456, "SeverityText":"WARN", "SeverityNumber":5, "TraceFlags":13, "Attributes":{"log.file.name":"a.log"}

 If you add in TraceId and SpanId to his example above, those fields dont get set still.  You can test with this and see.
 
 Metadata {"ObservedTimestamp":1234, "Timestamp":3456, "TraceId": "95e1d11ece6460e7d00c61d45cc195ff", "SeverityText":"WARN", "SeverityNumber":5, "TraceFlags":13, "Attributes":{"log.file.name":"a.log"}

**Comment by nokute78** [comment]
@cb645j The type of TraceId and SpanId are byte sequence.
String will be ignored since the data type is different.

https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-traceid
https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-spanid

**Comment by cb645j** [comment]
@nokute78 thanks, yes i figured that out.  For when getting from message of log event, since the traceid and spanid are strings in this case, i had to convert to byte arrays.  Please see my PR and make any suggestions or comments.  i was able to get it working for the 4 fields i added (span-id, trace-id, sev text, and sev number).

https://github.com/fluent/fluent-bit/pull/8576

File: plugins/out_opentelemetry/opentelemetry.h
Comment (r0mdau) [inline plugins/out_opentelemetry/opentelemetry.h:88]: @nokute78 what do you intended to do with this property ?

There is a need for setting OTLP resource attributes https://github.com/fluent/fluent-bit/issues/7649

**Pull request merged**
